### ISSUE TITLE
Getting Pod and Container Start/End Times to millisecond level precision

### ISSUE THREAD

**Issue by pritishnahar95** [kubernetes/kubernetes#81026]
**What would you like to be added**: Timestamps for Pod and Container Start/End times as well as events at a millisecond precision level.  

**Why is this needed**:
This is needed because I am trying to look into the Kubernetes overhead at a millisecond level precision and design my system around it.

**Comment by mickours** [comment]
We have the same need for experiments around Kubernetes scheduling. It would be great to have milliseconds on all events timestamps.

**Comment by lavalamp** [comment]
first time I saw this, sorry--

The time code is here: https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/time.go

There is a micro_time.go file right next to it.

One could attempt to either: change time.go to store milliseconds, or add a milli_time.go and then change the desired uses in the API.

Either way, there's a significant backwards compatibility problem-- you have to somehow show that all old clients continue to work fine when ms data is present. I don't know how you could do that. @liggitt may have an idea but will probably say the same thing.

**Comment by lavalamp** [comment]
(reopening this only because it should get a yes/no decision rather than aging out)

(It's looking like a 'no' IMO)

**Comment by liggitt** [comment]
> Either way, there's a significant backwards compatibility problem-- you have to somehow show that all old clients continue to work fine when ms data is present. I don't know how you could do that.

Agreed, changing the serialized format of that field is not backwards compatible.

**Comment by alculquicondor** [comment]
is there a v2 backlog? :)
In the meantime, what about adding a parallel field in v1? Something like `creationTimestampMs`?

**Comment by lavalamp** [comment]
> Agreed, changing the serialized format of that field is not backwards compatible.

Yeah, let's close it.

Parallel fields are a ton of work, I think we would need a lot more justification before doing something like that.

**Comment by sgarg04** [comment]
@pritishnahar95 I am also trying to do the same for pod and node events, Is there any resolution (or a way ) yet?

**Comment by andrejusc** [comment]
If there is such big backward compatibility concern here and due to that issue was closed - would it be possible then to adjust API and accept client's request for precision?

i.e. something like:
````
kubectl describe pod xyz --precision millis
````

Then old clients could work as is and whoever needs and could parse/handle different precision format will benefit from such.

**Comment by alculquicondor** [comment]
At a quick glance, that feasible to me, as part of GetOptions in the client.

But after digging into the implementation details, it's unclear how we can change the types to be able to change behavior per request https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/time.go

**Comment by alculquicondor** [comment]
Another thing to think through is what happens during an upgrade, where apiservers might be in different versions.

We are very late in the 1.27 release and this sounds worthy of a KEP in 1.28 or later. But probably best to bring the topic to a SIG API Machinery meeting https://github.com/kubernetes/community/tree/master/sig-api-machinery

**Comment by andrejusc** [comment]
> But after digging into the implementation details, it's unclear how we can change the types to be able to change behavior per request https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/time.go

Ii'm not familiar which flows call/instantiate corresponding Time objects, but it could have some optional flag as part of structure and depending on that flag presence/value - some other conditional logic for formatting needs then could happen.

I'm dealing now with pipelines, which spin a lot of involved pods and I'd like to be able to say till milliseconds what is the diff between creation and completion of some job/task/step/etc. So there is some sense of urgency to have such functionality available OOTB.

**Comment by liggitt** [comment]
> it's unclear how we can change the types to be able to change behavior per request

I don't think we would ever modify existing type serialization of a field that way... the expectation is that every serialized Pod object can be successfully decoded by every client every linked to Pod API types.

A parallel field is the only way I could see this working, and I agree with @lavalamp that that's a *lot* of work, has weird / problematic edge cases involving client incompatibility and data loss, and it's unclear that effort / risk is justified here.

**Comment by andrejusc** [comment]
>  the expectation is that every serialized Pod object can be successfully decoded by every client

I think it's up to client to provide a hint what extent of de-serilization it supports. And by using or not explicit additional options in request to indicate if it cares or not about backward compatibility. Various custom JSON mappers for serialization/de-serialization is nothing new in general. Then it mostly down to which specific schema is capable or not to validate and allow to parse some outcome.

Alignment to just seconds precision these days seems to be a bit outdated and makes it harder to correlates to other logs/events/etc.

**Comment by lavalamp** [comment]
The system doesn't work like that, we can't have pods that break (de)serialization, that will break controllers and custom extensions, none of which have a way of indicating they can't understand certain things, and it's also not optional to run these things, a cluster would be non-functional.

The precision of the time field is part of the type schema. We can't adjust the schema on a per-object basis, nothing in the system is prepared to handle that.

**Comment by alculquicondor** [comment]
Let's say we have an apiserver that can understand higher precision.
Could we have an extra layer in the endpoints (not serialization) that can lower the precision to seconds by default? And if there is a GetOption for higher precision, send that precision?

I know, very high level, but it sounds to me that it would be a backwards compatible approach.

**Comment by lavalamp** [comment]
That would require the higher precision be stored for everything already, which means kubelets and some controllers have to participate. It's actually harder to do that imo than to either migrate everybody or make a parallel field. The work required is just way out of line with the possible benefit.

**Comment by andrejusc** [comment]
Reading more on this topic and current API conventions in regards to serialization here: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#serialization-format
states that:

> All dates should be serialized as RFC3339 strings.

and then looking at mentioned RFC: https://www.rfc-editor.org/rfc/rfc3339 - it doesn't have mentioning of milliseconds. Hmm, I thought it would be there.

Then at least from current API conventions - not much to do.

Though same part of conventions has this point:

> APIs may return alternative representations of any resource in response to an Accept header or under alternative endpoints

which is what I'm potentially alluding here to. I.e. alternative representation is doable assuming that server-side has stored higher precision for some resource's fields (aka Pod) in question.

Then if current latest version of Pod spec/schema is v1 - I'd vote for v2 introduction to support higher precision and then migration where/if needed. Parallel field would make just more noise in already long jsons/yamls.

**Comment by lavalamp** [comment]
The writes of these time values are also done via the api. The read path is at best half the problem.

**Comment by andrejusc** [comment]
Yeah, tried that with Go code and while my current time in output is represented like this, with not just millis but I think nanos:
````
2023-03-02 21:03:29.634597818 +0000 GMT m=+0.008504534
````
The one I get via k8s Go client and from such code:
````
fmt.Println(pod.Status.StartTime)
````
is like this:
````
2023-03-02 20:52:55 +0000 GMT
````
Personally I'd not think that k8s saves timestamps internally just to second alignment. But it looks that way like you've said - then it's bigger design/implementation limitation IMO. Strange that having this Issue opened originally in 2019 - such was not addressed till now.

**Comment by lavalamp** [comment]
The API type is the storage type. Extra precision was seen as just making things bigger without adding value when we (quite possibly me personally) first added time fields to the api. API objects were much smaller then.

Fixing this now is too much effort compared to the value it would provide, so it will continue to be this way for the foreseeable future.

**Comment by andrejusc** [comment]
You mean you wanted to win on things like such:
https://github.com/kubernetes/api/blob/master/core/v1/types.go#L4124
i.e. 7 bytes instead of let say 10 within gRPC protobufs? 
And just possibly because of speed of client/server API interaction - such sacrifice to precision was made? 

It's a pain when internal Data object representation is tightly coupled to potential direct serialization without any optional transformation hook in between. Cutting 10 bytes to another 7 bytes would not be big deal from performance and before sending back response.

I mean - if platform (aka k8s) is evolving - then API should be capable to evolve as well. You may know better how additional extra bytes would multiply at the end per storage for some cluster of 1K+ resources. But probably k8s storage constraints evolved as well since 2019 and maybe it's time to reconsider this issue as well, which I'd really appreciate. And much more appeared in the space of extensions/CRDs, which blew up storage limits anyway I'd imagine.

Could we reopen this please or do we need to vote/bring to API machinery meeting as was suggested earlier today?

**Comment by lavalamp** [comment]
There wasn't a gRPC proto format when we added Time. gRPC might not have existed as a thing at that time.

Like I said, this is no longer about the size. It is about the compatibility problems. Fixing this now is too much effort compared to the value it would provide.

We can talk about this at a SIG meeting if you really want to, but I will just be repeating the things I've said here and the answer isn't going to change (sorry).

**Comment by andrejusc** [comment]
Just as a sake of experiment I did such:
````
kubectl describe pod xyz -o yaml > pod.yaml
````
Then inside that resulting pod.yaml I've changed this place for some of container inside that pod:
````
startedAt: "2023-03-02T23:17:47Z"
````
to this, i.e. adding as milliseconds 123 at the end via dot separation:
````
startedAt: "2023-03-02T23:17:47.123Z"
````
And then I've applied such via patch like this:
````
kubectl patch pod xyz --patch-file pod.yaml
pod/xyz patched (no change)
````
which means by itself that parsing works without an issue.

Now if instead I change to incorrect notation like this using extra colon:
````
startedAt: "2023-03-02T23:17:47:123Z"
````
then patch command fails as expected on such:
````
parsing time "2023-03-02T23:17:47:123Z" as "2006-01-02T15:04:05Z07:00": cannot parse ":123Z" as "Z07:00"
````

So at least to me above result means that input could contain millis and that will not affect at least patching case.

Would be than of value to know in more details what not storage but compatibility issues we're talking about here.

**Comment by andrejusc** [comment]
@lavalamp - was my above experiment of any use and to clarify what sort of compatibility we may have if not focusing on storage part?

**Comment by lavalamp** [comment]
Not really. kubectl probably fixed the file before sending it to apiserver. use --v=9 to see what was actually sent. or perhaps apiserver fixed it.

But this experiment is actually not telling us anything because the problematic thing is when apiserver returns extra precision to clients that don't expect it.

Another problem is storing extra precision and then downgrading to a version of apiserver that doesn't expect it.

**Comment by alculquicondor** [comment]
> Another problem is storing extra precision and then downgrading to a version of apiserver that doesn't expect it.

This risk can be reduced if we do it across two versions:
- version N+1 of apiserver knows how to parse and keep the extra precision from storage, but drops the precision in new objects.
- version N+2 stores new objects with higher precision.

**Comment by andrejusc** [comment]
Using `--v=9` I have as part of output such:

````
I0307 18:56:51.638906  231000 round_trippers.go:577] Response Headers:
I0307 18:56:51.638915  231000 round_trippers.go:580]     Date: Tue, 07 Mar 2023 18:56:56 GMT
I0307 18:56:51.638921  231000 round_trippers.go:580]     Audit-Id: 33008a4c-5059-48db-b0f8-97d4430c3893
I0307 18:56:51.638927  231000 round_trippers.go:580]     Cache-Control: no-cache, private
I0307 18:56:51.638933  231000 round_trippers.go:580]     Content-Type: application/json
I0307 18:56:51.638939  231000 round_trippers.go:580]     Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
I0307 18:56:51.638946  231000 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: f2cc461c-0591-406f-9684-340246cda147
I0307 18:56:51.638951  231000 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 8b60a7c7-5234-4a94-84a1-4937fb96b845
I0307 18:56:51.742739  231000 request.go:1154] Response Body:
...
"state":{"running":{"startedAt":"2023-03-07T18:55:05.123Z"}}}
...
[truncated 5532 chars]
````

Mine `kubectl `is at the moment:
```
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.3", GitCommit:"434bfd82814af038ad94d62ebe59b133fcb50506", GitTreeState:"clean", BuildDate:"2022-10-12T10:57:26Z", GoVersion:"go1.19.2", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.4+77bec7a", GitCommit:"b6d1f054747e9886f61dd85316deac3415e2726f", GitTreeState:"clean", BuildDate:"2022-12-14T20:18:42Z", GoVersion:"go1.19.4", Compiler:"gc", Platform:"linux/amd64"}
````

Does this help?

**Comment by lavalamp** [comment]
It shows that apiserver is truncating the time (you've elided lines wrong btw, that's part of a request not a response), but as I said, that doesn't address any of the problems

**Comment by lavalamp** [comment]
@alculquicondor yeah, I didn't say that problem is insolvable. The one where we don't know if clients can tolerate more detail is closer to being insolvable

Again, I'm not saying this is impossible, just that the effort drastically exceeds the value

**Comment by alculquicondor** [comment]
If I recall correctly, my initial motivation to get involved in this thread was for better granularity in clusterloader2 results, which uses API fields for some measurements.

Some more precise results can be obtained from metrics, but given that metrics can't be per object (for example, per pod), it's not possible to combine pods with different characteristics in a load test and partition the metrics for each.

In other words, from SIG Scheduling I would welcome this change.
But I acknowledge that maintaining backwards compatibility adds a lot of design and implementation complexity to this feature request. So I don't expect the maintainers to solve this problem, but I hope we can still assist a contributor that is willing to do all the work. Granted, that the contributor needs to be highly knowledgeable about apiserver already, otherwise the contributor will require a lot of guidance and pretty much be a similar effort for the maintainers to explain what needs to be done compared to doing it themselves.

**Comment by andrejusc** [comment]
My motivation at the moment for this to be more granular is to be able to calculate till millisecond overall invocation of pods involved into Tekton pipelines and tasks in them. And I could only envision that adoption of those new k8s based pipelines will only grow, so this is then a blocker for that wide adoption.

**Comment by vishnuchalla** [comment]
Just trying to understand. @lavalamp, Are these the only two places where we need to make changes while we store the timestamps?
* For all the references of these methods: https://github.com/kubernetes/kubernetes/blob/cc2f7b319801e1fd983a09e2442001e44010827b/staging/src/k8s.io/apimachinery/pkg/api/meta/meta.go#L550C26-L550C46

And for pod statuses

* https://github.com/kubernetes/kubernetes/blob/cc2f7b319801e1fd983a09e2442001e44010827b/pkg/kubelet/status/status_manager.go#L724

And also is it guaranteed that the `PodScheduled` time is always greater the Pod `CreationTimestamp`?
https://groups.google.com/g/kubernetes-dev/c/rUniUqhI5YM

**Comment by ad-zsolt-imre** [comment]
It'd be great if we had event timestamps in milliseconds. I understand "the effort drastically exceeds the value" but for us, it'd be very valuable. Yes, we do some automation based on events and while seconds work, they are problematic. (I admit events being kinda considered as "best effort" and not guaranteed is a lot more problematic)

**Comment by linxiulei** [comment]
Passing by...

the precise timestamp is actually in the event name 

https://github.com/kubernetes/client-go/blob/v0.33.1/tools/record/event.go#L490-L493

	return &v1.Event{
		ObjectMeta: metav1.ObjectMeta{
			Name:        util.GenerateEventName(ref.Name, t.UnixNano()),
			Namespace:   namespace,


So you can reverse the event name to the time in UnixNano

**Comment by igooch** [comment]
What's the feasibility at this point in time for re-opening this issue?

As Kubernetes increasingly powers 'instant-on' workloads (such as our [agent-sandbox](https://github.com/kubernetes-sigs/agent-sandbox) project), the current 1-second precision of `creationTimestamp` has become a primary bottleneck for accurate performance tracking and autoscaling.

**The Data: High-Precision vs. Standard Timestamps**
In our recent benchmarks, we compared startup latency using the built-in `creationTimestamp` (1-second precision) against a high-precision millisecond timestamp injected at the admission phase via a mutating webhook. The results show that the built-in metrics are skewed by over 500 ms at the median and over 1.5 seconds at the tail:

Metric  |  Standard K8s Timestamp (Seconds)  |  Admission-Time Annotation (ms)  |  Delta (Error)
------- | -------- | ------- | --------
P50 Latency  |  748.54 ms  |  215.19 ms  |  533.35 ms
P90 Latency  |  1796.88 ms  |  451.03 ms  |  1345.85 ms
P99 Latency  |  2429.69 ms  |  863.64 ms  |  1566.05 ms

We have explored several workarounds, but none are suitable for standard production use:

Mutating Admission Webhooks: These allow for `time.Now()` injection in Go, but they introduce operational fragility, certificate management overhead, and increased API server latency.

Mutating Admission Policy (MAP): While MAP is the preferred declarative alternative to webhooks, it is not currently an option because Kubernetes CEL does not provide a now variable or equivalent function to retrieve the current time during the admission request.
